uber/templates/macros.html

```diff
@@ -21,46 +21,205 @@
+          class="checkbox-label{% if label_class %} {{ label_class }}{% endif %}{% if is_readonly %} disabled{% endif %}"
+          {% if title %}title="{{ title }}"{% endif %}>
     {% endif %}
-    {% if readonly %}
+    {% if is_readonly %}
       <input
-        type="hidden"
-        name="{{ field_name }}"
-        id="{% if id %}{{ id }}{% else %}{{ field_name }}{% endif %}"
-        value="{%- if model[field_name] -%}1{%- else -%}0{%- endif -%}" />
+          type="hidden"
+          name="{{ field_name }}"
+          id="{% if id %}{{ id }}{% else %}{{ field_name }}{% endif %}"
+          value="{%- if model[field_name] -%}1{%- else -%}0{%- endif -%}" />
       <input
-        type="checkbox"
-        value="1"
-        disabled
-        {% if not label and title %}title="{{ title }}"{% endif %}
-        {% if model[field_name] %}checked{% endif %} />
+          type="checkbox"
+          value="1"
+          disabled
+          {% if not label and title %}title="{{ title }}"{% endif %}
+          {% if model[field_name] %}checked{% endif %} />
     {% else %}
       <input
-        type="checkbox"
-        name="{{ field_name }}"
-        id="{% if id %}{{ id }}{% else %}{{ field_name }}{% endif %}"
-        value="1"
-        {% if not label and title %}title="{{ title }}"{% endif %}
-        {% if model[field_name] %}checked{% endif %} />
+          type="checkbox"
+          name="{{ field_name }}"
+          id="{% if id %}{{ id }}{% else %}{{ field_name }}{% endif %}"
+          value="1"
+          {% if not label and title %}title="{{ title }}"{% endif %}
+          {% if model[field_name] %}checked{% endif %} />
     {% endif %}
     {% if label %}{{ label }}</label>{% endif %}
 {%- endmacro %}
 
 
-{% macro checkgroup(model, field_name, defaults=False, no_instance=False) -%}
-    {%- set defaults = model[field_name] if not defaults and not no_instance else None -%}
-    {%- set defaults = defaults.split(",") if defaults else [] -%}
+{% macro checkgroup(model, field_name, suffix='', defaults=False, is_readonly=False) -%}
+  {%- set is_class = model is class -%}
+  {%- set defaults = model[field_name] if not defaults and not is_class else '' -%}
+  {%- set defaults = defaults.split(",") if defaults else [] -%}
+  <div class="checkgroup">
     {% for num, desc in model.get_field(field_name).type.choices -%}
-        <label class="checkgroup checkbox-label" style="font-weight: normal;">
-            <input type="checkbox" name="{{ field_name }}" value="{{ num }}" {% if num|string in defaults %}checked{% endif %} />
+      {%- set has_value = num|string in defaults -%}
+      {%- if not is_readonly or has_value -%}
+        <label
+            class="checkbox-label{% if is_readonly %} disabled{% endif %}"
+            style="font-weight: normal;">
+          {% if is_readonly %}
+            {% if has_value %}
+              <input type="hidden" name="{{ field_name }}{{ suffix }}" value="{{ num }}"/>
```

Comment:
So, I think we've had this discussion before? If a checkbox is marked as `readonly` then you can still check/uncheck it – you just can't change it's `value` attribute. Which doesn't help us at all!

So that's why we use `disabled` to prevent changing it's checked/unchecked state. However! A `disabled` checkbox will NOT be included in the POST variables when you submit a form. Thus, to also include the correct variables we have to use a `hidden` input field.